File: apps/knowledge-service/src/services/doc-searcher.ts
import { readdir, readFile, stat } from "node:fs/promises";
import { join, extname } from "node:path";

export interface SearchResult {
  path: string;
  relevanceScore: number;
  excerpt: string;
  matchedKeywords: string[];
}

// ---------------------------------------------------------------------------
// Walk a directory tree collecting .md file paths
// ---------------------------------------------------------------------------
async function walkMd(dir: string): Promise<string[]> {
  const results: string[] = [];
  try {
    const entries = await readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = join(dir, entry.name);
      if (entry.isDirectory()) {
        // skip node_modules / hidden dirs
        if (entry.name.startsWith(".") || entry.name === "node_modules")
          continue;
        results.push(...(await walkMd(fullPath)));
      } else if (extname(entry.name) === ".md") {
        results.push(fullPath);
      }
    }
  } catch {
    // directory may not exist — skip
  }
  return results;
}

// ---------------------------------------------------------------------------
// Search docs by keyword matching
// ---------------------------------------------------------------------------
export async function searchDocs(
  keywords: string[],
  searchPaths: string[],
  maxResults = 5,
  maxTotalChars = 2000,
): Promise<SearchResult[]> {
  const allFiles: string[] = [];

  for (const searchPath of searchPaths) {
    try {
      const info = await stat(searchPath);
      if (info.isDirectory()) {
        allFiles.push(...(await walkMd(searchPath)));
      } else if (extname(searchPath) === ".md") {
        allFiles.push(searchPath);
      }
    } catch {
      // path doesn't exist — skip
    }
  }

  const scored: SearchResult[] = [];

  for (const filePath of allFiles) {
    try {
      const content = await readFile(filePath, "utf-8");
      const lower = content.toLowerCase();
      const matched: string[] = [];

      for (const kw of keywords) {
        if (lower.includes(kw.toLowerCase())) {
          matched.push(kw);
        }
      }

      // Must match at least 2 keywords to qualify
      if (matched.length < 2) continue;

      const excerpt = content.slice(0, 400);

      scored.push({
        path: filePath,
        relevanceScore: matched.length,
        excerpt,
        matchedKeywords: matched,
      });
    } catch {
      // unreadable file — skip
    }
  }

  // Sort by relevance desc, take top N
  scored.sort((a, b) => b.relevanceScore - a.relevanceScore);
  const topResults = scored.slice(0, maxResults);

  // Trim total excerpt chars to maxTotalChars
  let totalChars = 0;
  const trimmed: SearchResult[] = [];
  for (const r of topResults) {
    const remaining = maxTotalChars - totalChars;
    if (remaining <= 0) break;
    if (r.excerpt.length > remaining) {
      trimmed.push({ ...r, excerpt: r.excerpt.slice(0, remaining) });
      totalChars += remaining;
    } else {
      trimmed.push(r);
      totalChars += r.excerpt.length;
    }
  }

  return trimmed;
}

File: knowledge/langgraph/knowledge/services/doc_searcher.py
"""Project-doc search by keyword matching -- ported from doc-searcher.ts."""

from __future__ import annotations

import os
from pathlib import Path

from langgraph.knowledge.types import SearchResult


def _walk_md(directory: str) -> list[str]:
    """Recursively collect ``.md`` file paths, skipping hidden dirs and node_modules."""

    results: list[str] = []
    try:
        for entry in os.scandir(directory):
            if entry.is_dir(follow_symlinks=False):
                if entry.name.startswith(".") or entry.name == "node_modules":
                    continue
                results.extend(_walk_md(entry.path))
            elif entry.is_file() and entry.name.endswith(".md"):
                results.append(entry.path)
    except OSError:
        pass
    return results


async def search_docs(
    keywords: list[str],
    search_paths: list[str],
    max_results: int = 5,
    max_total_chars: int = 2000,
) -> list[SearchResult]:
    """Search markdown docs for keyword matches across *search_paths*."""

    all_files: list[str] = []
    for sp in search_paths:
        p = Path(sp)
        try:
            if p.is_dir():
                all_files.extend(_walk_md(sp))
            elif p.is_file() and p.suffix == ".md":
                all_files.append(sp)
        except OSError:
            continue

    scored: list[SearchResult] = []
    for file_path in all_files:
        try:
            content = Path(file_path).read_text(encoding="utf-8")
        except OSError:
            continue

        lower = content.lower()
        matched = [kw for kw in keywords if kw.lower() in lower]

        # Must match at least 2 keywords to qualify
        if len(matched) < 2:
            continue

        scored.append(
            SearchResult(
                path=file_path,
                relevance_score=len(matched),
                excerpt=content[:400],
                matched_keywords=matched,
            )
        )

    scored.sort(key=lambda r: r.relevance_score, reverse=True)
    top = scored[:max_results]

    # Trim total excerpt chars
    trimmed: list[SearchResult] = []
    total_chars = 0
    for r in top:
        remaining = max_total_chars - total_chars
        if remaining <= 0:
            break
        if len(r.excerpt) > remaining:
            trimmed.append(
                SearchResult(
                    path=r.path,
                    relevance_score=r.relevance_score,
                    excerpt=r.excerpt[:remaining],
                    matched_keywords=r.matched_keywords,
                )
            )
            total_chars += remaining
        else:
            trimmed.append(r)
            total_chars += len(r.excerpt)

    return trimmed
